galley/pkg/config/processing/snapshotter/snapshotter.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snapshotter
+
+import (
+	"time"
+
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/processing/snapshotter/strategy"
+	"istio.io/istio/galley/pkg/runtime/monitoring"
+)
+
+// Snapshotter is a processor that handles input events and creates snapshot collections.
+type Snapshotter struct {
+	accumulators map[collection.Name]*accumulator
+	selector     *event.Selector
+	xforms       []event.Transformer
+	options      []SnapshotOption
+
+	// lastEventTime records the last time an event was received.
+	lastEventTime time.Time
+
+	// pendingEvents counts the number of events awaiting publishing.
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshot was published.
+	lastSnapshotTime time.Time
+}
+
+var _ event.Processor = &Snapshotter{}
+
+// HandlerFn handles generated snapshots
+type HandlerFn func(*collection.Set)
+
+type accumulator struct {
+	reqSyncCount int
+	syncCount    int
+	collection   *collection.Instance
+	strategies   []strategy.Instance
+}
+
+// Handle implements event.Handler
+func (a *accumulator) Handle(e event.Event) {
+	switch e.Kind {
+	case event.Added, event.Updated:
+		a.collection.Set(e.Entry)
+		monitoring.RecordStateTypeCount(e.Source.String(), a.collection.Size())
+	case event.Deleted:
+		a.collection.Remove(e.Entry.Metadata.Name)
+		monitoring.RecordStateTypeCount(e.Source.String(), a.collection.Size())
+	case event.FullSync:
+		a.syncCount++
+	default:
+		scope.Errorf("accumulator.Handle: unhandled event type: %v", e.Kind)
+		return
+	}
+
+	if a.syncCount >= a.reqSyncCount {
+		for _, s := range a.strategies {
+			s.OnChange()
+		}
+	}
+}
+
+func (a *accumulator) reset() {
+	a.syncCount = 0
+	a.collection.Clear()
+}
+
+// NewSnapshotter returns a new Snapshotter.
+func NewSnapshotter(xforms []event.Transformer, options []SnapshotOption) *Snapshotter {
```

Comment:
Why would a snapshotter perform transformations on the data? Should other processors do that?

Reply:
Snapshotter is a Processor. It is possible to create non-snapshotting Processors (which could be pretty useful for transferring streaming stuff, in the future).

We can create an intermediate layer for encapsulating transformations, but I wanted to avoid the complication before we truly need it.